### PR TITLE
Improve the Quantum Storage Renderer

### PR DESCRIPTION
**Pull request by ghzdude**
## What
This pr improves the quantum storage render by rendering the minimum amount of faces needed (15). Previously, it would render 31 faces.

also improves quantum tank rendering fluid contents by only rendering the two needed faces instead of all six

## Implementation Details
idk why a `foreach()` was used for the `boxFacingMap`
also previously it was `facing, boxFacingMap.get(box)` in `renderFace()`, which looked backwards to me
store the TextTexture as a static final variable instead of instantiating it every frame

## Outcome
more fps

## Potential Compatibility Issues
none

### CODE REVIEW

**Comment by ghzdude** [comment]
Here's a sparkc profile [without the renderer changes](<https://spark.lucko.me/zTL092CA7D>) (in the quantum controller pr brance), and here's [with the render changes](<https://spark.lucko.me/IA84Lq2EiY>). I'm not sure how accurate these are, but they were profiled for 300s while I'm looking at ~100 quantum storage chests and not moving.

2nd sparkc profile [without the changes](<https://spark.lucko.me/XErO0lDZZW>), and another [with the changes](<https://spark.lucko.me/ZdNlq8J4zt>)

**Comment by ghzdude** [comment]
(copied from discord)

New profile [with the text rendering removed + previous changes](<https://spark.lucko.me/0AWNJqvrOX>), and [without any changes](<https://spark.lucko.me/Fx3MISvG6e>) on master (with command `/sparkc profiler --timeout 300`)

**Comment by ghzdude** [comment]
New profiles:
[with changes](https://spark.lucko.me/44AP0A6fps)
without changes - cannot get a profile due to dev env issues

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java [hunk 1]
```diff
@@ -134,9 +135,17 @@ public void update() {
...
+                            int currentFill = MathHelper
+                                    .floor(16 * ((float) currentFluid.amount) / fluidTank.getCapacity());
+                            int previousFill = MathHelper
+                                    .floor(16 * ((float) previousFluid.amount) / fluidTank.getCapacity());
                             // tank has fluid with changed amount
                             previousFluid.amount = currentFluid.amount;
-                            writeCustomData(UPDATE_FLUID_AMOUNT, buf -> buf.writeInt(currentFluid.amount));
+                            writeCustomData(UPDATE_FLUID_AMOUNT, buf -> {
+                                buf.writeInt(currentFluid.amount);
+                                buf.writeBoolean(currentFill != previousFill);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java:146]: This doesn't seem to be used at all to space out rendering updates when reading the custom data

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java [hunk 2]
```diff
@@ -450,8 +460,8 @@ public void receiveCustomData(int dataId, PacketBuffer buf) {
             try {
                 this.fluidTank.setFluid(FluidStack.loadFluidStackFromNBT(buf.readCompoundTag()));
             } catch (IOException ignored) {
-                GTLog.logger.warn("Failed to load fluid from NBT in a quantum tank at {} on a routine fluid update",
-                        this.getPos());
+                GTLog.logger.warn("Failed to load fluid from NBT in a quantum tank at " + this.getPos() +
+                        " on a routine fluid update");
             }
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java:465]: This change does not seem necessary, since both do the same thing.

File: src/main/java/gregtech/client/renderer/texture/custom/QuantumStorageRenderer.java [hunk 3]
```diff
@@ -110,20 +138,23 @@ public static void renderChestStack(double x, double y, double z, MetaTileEntity
...
+
         float lastBrightnessX = OpenGlHelper.lastBrightnessX;
         float lastBrightnessY = OpenGlHelper.lastBrightnessY;
+
         if (world != null) {
             renderState.setBrightness(world, pos);
         }
-        FluidStack stack = tank.getFluid();
-        if (stack == null || stack.amount == 0)
-            return;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/client/renderer/texture/custom/QuantumStorageRenderer.java:121]: Why was this removed? It would seem useful to skip rendering if there is no fluid, or the fluid in null.

File: src/main/java/gregtech/api/gui/resources/TextTexture.java [hunk 4]
```diff
@@ -42,11 +52,18 @@ public TextTexture setDropShadow(boolean dropShadow) {
...
-        if (FMLCommonHandler.instance().getSide().isClient()) {
-            if (this.width > 0) {
-                texts = Minecraft.getMinecraft().fontRenderer.listFormattedStringToWidth(text, width);
-            } else {
-                texts = Collections.singletonList(text);
+        return this;
+    }
+
+    public TextTexture setText(String text) {
+        if (this.text.isEmpty() || !this.text.equals(text)) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/gui/resources/TextTexture.java:59]: `this.text.isEmpty()` skips the case where the incoming `text` parameter is also empty, in which case nothing needs to be set.

File: src/main/java/gregtech/client/renderer/texture/custom/QuantumStorageRenderer.java [hunk 5]
```diff
@@ -69,18 +69,21 @@ public void renderMachine(CCRenderState renderState, Matrix4 translation, IVerte
...
-            for (EnumFacing box : EnumFacing.VALUES) {
-                if ((facing != frontFacing || box != frontFacing) &&
-                        (facing != EnumFacing.DOWN || box.getAxis().isVertical())) { // Don't render the front facing
-                                                                                     // box from the front, nor allow
-                                                                                     // Z-fighting to occur on the
-                                                                                     // bottom
-                    Textures.renderFace(renderState, translation, pipeline, facing, boxFacingMap.get(box), hullTexture,
+        for (EnumFacing facing : EnumFacing.VALUES) {
+            for (EnumFacing boxFace : EnumFacing.VALUES) {
+                // do not render the box when "facing" is "frontFacing", otherwise
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/client/renderer/texture/custom/QuantumStorageRenderer.java]: the outer loop is for the sides of the machine, the inner loop is for rendering the boxes for the inside and outside of the chest. essentially rendering certain faces of a box for each side.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java (hunk 1 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java:146]: hmm, i might have messed something up when i was rebasing this

File: src/main/java/gregtech/client/renderer/texture/custom/QuantumStorageRenderer.java (hunk 3 above)
Comment (ghzdude) [reply src/main/java/gregtech/client/renderer/texture/custom/QuantumStorageRenderer.java:121]: the checks were moved up

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java (hunk 2 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumTank.java:465]: also probably a mistake with rebasing

File: src/main/java/gregtech/api/gui/resources/TextTexture.java (hunk 4 above)
Comment (ghzdude) [reply src/main/java/gregtech/api/gui/resources/TextTexture.java:59]: tbh the `isEmpty()` can be removed. the not equals check should be enough

**Review by ALongStringOfNumbers — APPROVED** [review]

**Review by ALongStringOfNumbers — APPROVED** [review]

**Review by ALongStringOfNumbers — APPROVED** [review]
